### PR TITLE
Feat/colour perception

### PR DESCRIPTION
**Pull request by kargaranamir**
Remaining `colour_perception` metrics from the legacy.

### CODE REVIEW

**Comment by kargaranamir** [comment]
- [x] Run evaluator for m20
- [x] Update metrics.json
- [x] Add comment about difference between luminances in m13, m14.

**Comment by kargaranamir** [comment]
- [x] Why these Luminance Types? Based on Legacy
- [x] Short Description in Judgments

**Comment by mplaine** [comment]
In the comments/documentation, the range for all the results of LAB avg and std (m14) have been defined as [0, +inf). However, some evaluation results include negative values.
-> @kargaranamir Please, double check and update the implementation and/or ranges.

**Comment by kargaranamir** [comment]
Thanks. I will check it ASAP.

**Comment by mplaine** [comment]
Could you add more tests to verify the results of the Colorfulness metric (m15)? In their paper, they state _"Since we consider that a greyscale image has no colourfulness [...]"_. However, some of our greyscale test images output results that are significantly higher than zero. Is this behavior correct?

**Comment by mplaine** [comment]
Should we add **Hue standard deviation** as m16_1?

**Comment by kargaranamir** [comment]
> Should we add **Hue standard deviation** as m16_1?

Hue is a circular quantity, so it's not obvious how to compute Hue STD. Here are the links we used for the AVG:
- https://en.wikipedia.org/wiki/Circular_mean
- http://mkweb.bcgsc.ca/color-summarizer/?faq#averagehue

Are you familiar with how to compute the circular quantity STD? Quick search didn't turn up any good links, since legacy doesn't support it either, I did not implement it.

**Comment by mplaine** [comment]
> > Should we add **Hue standard deviation** as m16_1?
> 
> Hue is a circular quantity, so it's not obvious how to compute Hue STD. Here are the links we used for the AVG:
> 
> * https://en.wikipedia.org/wiki/Circular_mean
> * http://mkweb.bcgsc.ca/color-summarizer/?faq#averagehue
> 
> Are you familiar with how to compute the circular quantity STD? Quick search didn't turn up any good links, since legacy doesn't support it either, I did not implement it.

I assumed something like this was the reason. Thanks for verifying. Let's leave the metrics as it is.

**Comment by kargaranamir** [comment]
> Could you add more tests to verify the results of the Colorfulness metric (m15)? In their paper, they state _"Since we consider that a greyscale image has no colourfulness [...]"_. However, some of our greyscale test images output results that are significantly higher than zero. Is this behavior correct?

I will design more test cases to see what happens. As a heads-up, we have implemented "7. A MORE EFFICIENT METRIC" from the paper. It reproduces the section 6 result more efficiently but it does not give the same result as section 6.

**Comment by mplaine** [comment]
> > Could you add more tests to verify the results of the Colorfulness metric (m15)? In their paper, they state _"Since we consider that a greyscale image has no colourfulness [...]"_. However, some of our greyscale test images output results that are significantly higher than zero. Is this behavior correct?
> 
> I will design more test cases to see what happens. As a heads-up, we have implemented "7. A MORE EFFICIENT METRIC" from the paper. It reproduces the section 6 result more efficiently but it does not give the same result as section 6.

Thanks for the clarification! Yes, M3.

**Comment by mplaine** [comment]
**m17:** According to the documentation, the max number of distinct HSV values is 255^3 (16581375). The generated histogram image, however, shows a max value of ~220.
@kargaranamir Is the implementation and/or documentation correct?

Ps. Should we implement 'Distinct RGB values' (m3) similarly, i.e., include the number of distinct Red, Green, and Blue values to the list of result measures as well?

**Comment by kargaranamir** [comment]
This is due to the paper threshold; only values covering more than 0.1% of image are counted, see "C1-C3". This is how I calculated 0.1 % of the image: `0.01 * image.width * image.height`, but I'm not sure if it's what the paper wanted. It is also unnecessary to count RGB in this way. This is because in HSV, the unique number of HSVs does not matter, but the distinct number of each channel matters.

**Comment by mplaine** [comment]
Aah, you are correct. I overlooked the 0.1% threshold filter and as all the results for distinct HSV were a tad under 255 I thought there could be a mistake in the code.

Thanks for pointing out the differences between RGB and HSV. C4 (m3) of the AVI'14 paper is then correctly implemented and there is no reason to count distinct values of Red, Green, and Blue. However, C1-C3 of the AVI'14 paper do not include a metric for the number of distinct HSV values (m17_0), right? Further, in your above comment you mentioned that _"the unique number of HSVs does not matter, but the distinct number of each channel matters"_.
-> Should m17_0 be removed if i) it does not make sense to compute it and ii) it is not included as a separate metric in the paper?

**Comment by kargaranamir** [comment]
Yes. I did not find any reference that explicitly pointed out the number of distinct HSV, but for the channels individually there are. I think we should delete m17_0, or if we want to keep it, there is no need to follow the 0.1% image threshold, cause this threshold is for the channels. 

Ps. We maybe need to rename "Distinct RGB values" or "Distinct HSV values" cause one is for the whole RGB and one is for each channel separately, and they are not the same way. However, it is acceptable to keep these names as well.

**Comment by mplaine** [comment]
Here my suggestion:
- Let's delete m17_0 and modify the implementation and metrics.json accordingly
- Let's rename m17 as _"Distinct values of Hue, Saturation, and Value"_

Could you kindly apply the above changes? I can then review the changes and run the evaluator + update the ranges in metrics.json.

**Comment by mplaine** [comment]
Hi Amir, please let me know when you are done with your changes / fixes so that I can re-review the metrics. I still have 'Color Harmonization' to be reviewed in this set of metrics.

**Comment by kargaranamir** [comment]
As soon as I make the changes, I will let you know. I have started doing a part now.

**Comment by kargaranamir** [comment]
- m14: I fixed m14 ranges, and also checked some test cases using this online converter: http://colormine.org/convert/rgb-to-lab
- m13, m15: I made a mistake in implementing m13 and m15, and instead of RGB, I used BGR.
- m15: I also made a mistake on the calculation of m15 based on the [legacy](https://github.com/aalto-ui/aim/blob/aim2/legacy/aim_metrics/aim_metrics/colour_perception/cp6_hassler_susstrunk.py). "abs" function should not be used, since it is not mentioned in the paper.  Based on what I found online, I checked my implementation with other implementations as well: [[1]](https://github.com/Collonville/ImageFeatureExtractor/blob/92c9b4bbb19ac6f319d86e2e9837425a822e78aa/ife/features/colourfulness.py), [[2]](https://github.com/stuisme/pix/blob/ea2a84ae52672742ca474ee9cf977b37469956bc/Losses/SaturationLoss.py), [[3]](https://github.com/DrNickRedfern/HSM3/blob/78ce5126988fe851eedb3678496818207770578c/HSM3.R)

All related files (metrics.json, ranges, implementation, histograms, tests, json values) are updated.

**Comment by kargaranamir** [comment]
> Here my suggestion:
> 
> * Let's delete m17_0 and modify the implementation and metrics.json accordingly
> * Let's rename m17 as _"Distinct values of Hue, Saturation, and Value"_
> 
> Could you kindly apply the above changes? I can then review the changes and run the evaluator + update the ranges in metrics.json.

Done.

**Comment by mplaine** [comment]
I reviewed metrics m13-m17. Only m20 to go.

**Comment by kargaranamir** [comment]
Great. I checked them again, and I am concerned about m17. It says only those values covering over 0.1% of the image were used. Does that mean HSV values covering over the threshold together or individually? I implemented the threshold for HSV values together, since it's a color variability threshold.

File: backend/aim/metrics/m17/m17_distinct_values_of_hue_saturation_and_value.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+        # Create lists to store values of HSV, Hue, Saturation, and Value from the histogram
+        hue_values: List[int] = []
+        saturation_values: List[int] = []
+        value_values: List[int] = []
+
+        # Collect distinct values of Hue, Saturation, and Value after color reduction
+        for hist in list(hsv_color_histogram):
+            hist_count, hist_value = hist
+            if hist_count > color_reduction_threshold:
```
Comment (kargaranamir) [inline backend/aim/metrics/m17/m17_distinct_values_of_hue_saturation_and_value.py]: m17 color reduction applied to hist_count, which is the number of HSV values.

Comment (mplaine) [reply backend/aim/metrics/m17/m17_distinct_values_of_hue_saturation_and_value.py]: That's indeed tricky to judge based on the papers alone!

RGB:
- C4 in AVI'14: _"The number of distinct **RGB values**. Color variability was reduced: only **colors** covering more than five pixels were counted."_
- A3 in NordiCHI'14: _"Following the procedure from [21], we counted distinct **pixel RGB values** met more than two times"_

HSV:
- C1-C3 in AVI'14: _"The number of distinct **values of Hue, Saturation and Value.** Images were converted to the HSV color space. Color variability was reduced: only **values** covering more than 0.1% of image were counted."_
- A5-A7 in NordiCHI'14: _"counted distinct **Hue, Saturation and Value pixel values** coveting more than 0.1% of screenshot"_

Q1: Have you tried to find other implementations or literature on how to count unique colors/values in RGB and HSV? For example, [https://www.mathworks.com/matlabcentral/answers/77025-count-colours-in-hsv](https://www.mathworks.com/matlabcentral/answers/77025-count-colours-in-hsv)
Q2: Which of the two approaches makes more sense, if any?

**Comment by mplaine** [comment]
Review of m20 is done, excluding checking m20_0 range values.

Do you still want to apply further changes to m17, or are we ready to merge this PR to the 'aim2' branch?

**Comment by kargaranamir** [comment]
I apologize for the late update. The m17 color reduction method has been changed to do it separately for each channel. I think this is more reasonable. If everything is okay with m20, we can merge.

**Comment by kargaranamir** [comment]
Have you checked m20's image outputs? Is everything okay with them?

**Comment by mplaine** [comment]
m20 image outputs looked fine. I will do a merge, run all metrics, check ranges, and close this PR.

**Pull request merged**